genai-perf/genai_perf/goodput_calculator/llm_goodput_calculator.py

```diff
@@ -0,0 +1,169 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from typing import Dict, Optional
+import genai_perf.logging as logging
+from genai_perf.goodput_calculator.goodput_calculator import GoodputCalculator
+from genai_perf.metrics.llm_metrics import LLMMetrics
+
+logger = logging.getLogger(__name__)
+
+class LLMGoodputCalculator(GoodputCalculator):
+    """
+    A subclass to calculate goodput for LLMs according to LLM-related SLOs.
+    """
+
+    def __init__(self,
+                goodput_constraints: Dict[str, float],
+                metric: LLMMetrics,
+                benchmark_duration: float,
+    ) -> None:
+        super().__init__(goodput_constraints, metric, benchmark_duration)
+        # (TMA-1975 related) The order is hardcoded as below due to the hardcoded order
+        # in LLMMetirc class. We would eventually want to impose some consistent order 
+        # for time-related metrics and throughput related metrics.
+        self._valid_time_related_names = [
+            item.name for item in metric.request_time_metrics
+        ]
+        self._valid_throughput_related_names = [
+            item.name for item in metric.request_throughput_metrics
+        ]
+        self._valid_metric_names = (
+            self._valid_time_related_names + self._valid_throughput_related_names
+        )
+        self._has_time_target = False
+        self._has_throughput_target = False
+
+    def _set_valid_slos(self) -> None:
+        """
+        Check users' Service Level Objectives (SLOs) inputs. 
+        Set the valid ones while logging the invalid ones. 
+        """
+        invalid_slos = []
+        self._valid_time_related_slos = {}
+        self._valid_throughput_related_slos = {}
+        for slo_name, slo_value in self._goodput_constraints.items():
+            try:
+                base_name = self._metric.get_base_name(slo_name)
+                if base_name in self._valid_metric_names:
+                    if base_name in self._valid_time_related_names:
+                        self._valid_time_related_slos[slo_name] = (
+                            slo_value * self.MS_TO_NS_CONVERSION
+                        )
+                    elif base_name in self._valid_throughput_related_names:
+                        self._valid_throughput_related_slos[slo_name] = (
+                            slo_value 
+                        )
+            except KeyError:            
+                invalid_slos.append(slo_name)
+        if self._valid_time_related_slos:
+            self._has_time_target = True
+        if self._valid_throughput_related_slos:
+            self._has_throughput_target = True 
+        if invalid_slos:
+            valid_slos_list = ', '.join(self._valid_metric_names)
+            logger.info(f"Invalid SLOs found: {', '.join(invalid_slos)}. "
+                        f"The goodput will be N/A. "
+                        f"Valid SLOs are: {valid_slos_list} in plural forms.")
```

Comment:
Is there a way to use singular forms? Like using get_base_name?

Reply:
Updated